swim/disseminator.go

```diff
@@ -245,3 +248,60 @@ func (d *disseminator) ChangesCount() int {
 	d.RUnlock()
 	return c
 }
+
+// tryStartReverseFullSync fires a goroutine that performs a full sync. We omit
+// the reverse full sync if the maximum number of processes are already
+// running. This ensures no more than reverseFullSyncJobs processes are
+// running concurrently.
+func (d *disseminator) tryStartReverseFullSync(target string, timeout time.Duration) {
+	if d.reverseFullSyncJobs == nil {
+		d.reverseFullSyncJobs = make(chan struct{}, d.node.maxReverseFullSyncJobs)
+	}
+
+	// occupy a job, return if none are available
+	select {
+	case d.reverseFullSyncJobs <- struct{}{}:
+		// continue if job is available
+	default:
+		d.logger.WithFields(log.Fields{
+			"remote": target,
+		}).Info("omit bidirectional full sync, too many already running")
+
+		d.node.emit(OmitReverseFullSyncEvent{Target: target})
+		return
+	}
+
+	// start reverse full sync
+	go func() {
+		d.reverseFullSync(target, timeout)
+
+		// create a new vacancy when the job is done
+		<-d.reverseFullSyncJobs
+	}()
```

Comment:
In this case, since there is little code, I agree. However, I've seen the situations where the case becomes very large and the code becomes impossible to follow. The lack of indentation in the form above is also a plus.

I've found that reducing the select statement into it's smallest form, is generally the easiest way to follow concurrent code.